staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/kms/kms.go

```diff
@@ -0,0 +1,177 @@
+const defaultCacheSize = 1000
+
+// Service allows encrypting and decrypting data using an external Key Management Service.
+type Service interface {
+	// Decrypt a given data string to obtain the original byte data.
+	Decrypt(data string) ([]byte, error)
+	// Encrypt bytes to a string ciphertext.
+	Encrypt(data []byte) (string, error)
+	// CheckStale checks if the provided encrypted text is stale and needs to be re-encrypted.
+	CheckStale(data string) (bool, error)
+}
+
+type kmsTransformer struct {
+	kmsService   Service
+	transformers *lru.Cache
+
+	// cacheSize is the maximum number of DEKs that are cached.
+	cacheSize int
+
+	lock sync.RWMutex
+}
+
+// NewKMSTransformer returns a transformer which implements a KEK-DEK based envelope encryption scheme.
+// It uses kmsService to encrypt and decrypt DEKs. Respective DEKs (in encrypted form) are prepended to
+// the data items they encrypt. A cache (of size cacheSize) is maintained to store the most recently
+// used decrypted DEKs in memory.
+func NewKMSTransformer(kmsService Service, cacheSize int) (value.Transformer, error) {
+	if cacheSize == 0 {
+		cacheSize = defaultCacheSize
+	}
+	cache, err := lru.New(cacheSize)
+	if err != nil {
+		return nil, err
+	}
+	return &kmsTransformer{
+		kmsService:   kmsService,
+		transformers: cache,
+		cacheSize:    cacheSize,
+	}, nil
+}
+
+// TransformFromStorage decrypts data encrypted by this transformer using envelope encryption.
+func (t *kmsTransformer) TransformFromStorage(data []byte, context value.Context) ([]byte, bool, error) {
+	fmt.Println("KMS T: FROM")
+	t.lock.RLock()
+	defer t.lock.RUnlock()
+
+	// Length of key can be encoded in 16 bits.
+	keyLen := int(binary.BigEndian.Uint16(data[:2]))
+	if keyLen+2 > len(data) {
+		return []byte{}, false, fmt.Errorf("invalid data encountered by gkms transformer, length longer than available bytes: %q", data)
+	}
+	encKey := string(data[2 : keyLen+2])
+	encData := data[2+keyLen:]
+
+	var transformer value.Transformer
+	var kmsStale bool
+	_transformer, found := t.transformers.Get(encKey)
+	if found {
+		transformer = _transformer.(value.Transformer)
+	} else {
+		key, err := t.kmsService.Decrypt(encKey)
+		if err != nil {
+			return []byte{}, false, fmt.Errorf("error while decrypting key: %q", err)
+		}
+
+		// We need to release the read lock to prevent a deadlock
+		t.lock.RUnlock()
+		transformer, err = t.addTransformer(encKey, key)
+		t.lock.RLock()
+
+		if err != nil {
+			return []byte{}, false, err
+		}
+	}
+	kmsStale, err := t.kmsService.CheckStale(encKey)
+	if err != nil {
+		return []byte{}, false, err
+	}
+	res, transformerStale, err := transformer.TransformFromStorage(encData, context)
+	return res, (transformerStale || kmsStale), err
+}
+
+// TransformToStorage encrypts data to be written to disk using envelope encryption.
+func (t *kmsTransformer) TransformToStorage(data []byte, context value.Context) ([]byte, error) {
+	fmt.Println("KMS T: TO")
+	newKey, err := generateKey(32)
```

Comment:
Can this key length be made configurable? With advances in GPU and CPU speeds, its a matter of time before longer key lengths are necessary.